### PR TITLE
SNS: added api for ListTopics

### PR DESCRIPTION
**Pull request by alexturkin**
### Problem

"ListTopics" action is absent.
### Solution

ADd "ListTopics" action to `erlcloud_sns` module

### CODE REVIEW

**Comment by alexturkin** [comment]
@aporubov-al please take a look

File: src/erlcloud_sns.erl [hunk 1]
```diff
@@ -276,7 +277,33 @@ list_endpoints_by_platform_application(PlatformApplicationArn, NextToken, Config
...
+-spec list_topics/1 :: (undefined | string() | aws_config()) -> [{topics, [[{arn, string()}]]} | {next_token, string()}].
+-spec list_topics/2 :: (undefined | string(), aws_config()) ->  [{topics, [[{arn, string()}]]} | {next_token, string()}].
 
+list_topics() ->
+    list_topics(default_config()).
+
+list_topics(Config) when is_record(Config, aws_config) ->
+    list_topics(undefined, Config);
+list_topics(NextToken) ->
+    list_topics(NextToken, default_config).
```
Comment (kkuzmin) [inline src/erlcloud_sns.erl]: I suppose `default_config()` call instead of atom here.

**Comment by kkuzmin** [comment]
+1 other than the comment above

File: src/erlcloud_sns.erl [hunk 2]
```diff
@@ -276,7 +277,33 @@ list_endpoints_by_platform_application(PlatformApplicationArn, NextToken, Config
...
     list_endpoints_by_platform_application(PlatformApplicationArn, NextToken, new_config(AccessKeyID, SecretAccessKey)).
 
+-spec list_topics/0 :: () -> [{topics, [[{arn, string()}]]} | {next_token, string()}].
+-spec list_topics/1 :: (undefined | string() | aws_config()) -> [{topics, [[{arn, string()}]]} | {next_token, string()}].
+-spec list_topics/2 :: (undefined | string(), aws_config()) ->  [{topics, [[{arn, string()}]]} | {next_token, string()}].
 
+list_topics() ->
+    list_topics(default_config()).
+
+list_topics(Config) when is_record(Config, aws_config) ->
```
Comment (aporubov-al) [inline src/erlcloud_sns.erl:288]: minor: why not list_topics(#aws_config{} = Config) -> ... ?

File: src/erlcloud_sns.erl [hunk 3]
```diff
@@ -276,7 +277,33 @@ list_endpoints_by_platform_application(PlatformApplicationArn, NextToken, Config
...
+        end,
+    Doc = sns_xml_request(Config, "ListTopics", Params),
+    Decoded =
+        erlcloud_xml:decode(
+            [{topics, "ListTopicsResult/Topics/member",
+                fun extract_topic_arn/1
+             },
+             {next_token, "ListTopicsResult/NextToken", text}],
+            Doc),
+    Decoded.
```
Comment (aporubov-al) [inline src/erlcloud_sns.erl:307]: maybe it makes sense to provide a version of this function which iterates over "NextTokens" automatically effectively retrieving all topics without making a client function manually call list_topics(NextToken, Config), there is an example of that in erlcloud_rds

File: src/erlcloud_sns.erl [hunk 4]
```diff
@@ -571,6 +598,9 @@ extract_attribute(Nodes) ->
       erlcloud_xml:get_text("value", Node)}
      || Node <- Nodes].
 
+extract_topic_arn(Nodes) ->
+    [erlcloud_xml:decode([{arn, "TopicArn", text}], Node) || Node <- Nodes].
```
Comment (aporubov-al) [inline src/erlcloud_sns.erl:621]: Maybe that's better?

```
extract_topic_arn(Nodes) when is_list(Nodes) ->
    [extract_topic_arn(Node) || Node <- Nodes];
extract_topic_arn(Node) ->
    [{arn, erlcloud_xml:get_text("TopicArn", Node)}].
```

Just to clearly separate a "list" and a "single element" cases. But that's a really minor thing, it's up to you to stay with your current solution or pick up this one

File: src/erlcloud_sns.erl (hunk 3 above)
Comment (alexturkin) [inline src/erlcloud_sns.erl:307]: I like this idea! I'll add it

**Comment by nalundgaard** [comment]
Please consider making this to gleber. When you add primary stuff like this in our fork, you risk someone else implementing it slightly differently in the mainline, and then us having a mess down the road. This has already happened with erlcloud_ddb2. I did this with erlcloud_ddb_streams with @smilerlee, and it was a good experience. It doesn't take that much time to get it into the mainline and then merge down here.

File: src/erlcloud_sns.erl (hunk 2 above)
Comment (alexturkin) [inline src/erlcloud_sns.erl:288]: `when is_record(Config, aws_config)` is used everywhere in that file, so I repeated that pattern

**Comment by alexturkin** [comment]
@nalundgaard great idea! 

@motobob Do you agree with making PR to the upstream, and then merge it to ours?

If yes, I'll close this PR and make another to gleber repo. (Branch will be still there, so we can use it before it'll be merged)

File: src/erlcloud_sns.erl (hunk 4 above)
Comment (alexturkin) [inline src/erlcloud_sns.erl:621]: Yes, it makes sense, but I decided to leave it as is

**Comment by smilerlee** [comment]
@alexturkin It would be much great if the remaining unimplemented APIs could also be added. Yet that'll be a lot of works to do.

**Comment by motobob** [comment]
@alexturkin i'll merge it here to unblock our work and later when this makes to upstream will FF to it

**Pull request merged**